mlflow/models/evaluation/default_evaluator.py

```diff
@@ -462,21 +477,70 @@ def _evaluate_custom_metric(custom_metric_tuple, eval_df, builtin_metrics):
     :param custom_metric_tuple: Containing a user provided function and its index in the
                                 ``custom_metrics`` parameter of ``mlflow.evaluate``
     :param eval_df: A Pandas dataframe object containing a prediction and a target column.
-    :param builtin_metrics: A dictionary of metrics produced by the default evaluator.
-    :return: A scalar metric value.
+    :param builtin_metrics: A dictionary of aggregate metrics produced by the default evaluator.
+    :param metrics: A dictionary of metric values produced by the default evaluator.
+    :return: MetricValue
     """
     exception_header = (
         f"Custom metric '{custom_metric_tuple.name}' at index {custom_metric_tuple.index}"
         " in the `custom_metrics` parameter"
     )
 
-    metric = custom_metric_tuple.function(eval_df, builtin_metrics)
+    metrics_type = get_type_hints(custom_metric_tuple.function).get("metrics")
+    if metrics_type and metrics_type == Dict[str, MetricValue]:
```

Comment:
Curious why we need this check. Can we just check the return value?

Reply:
I think we want to ask users to provide the type hint in order to use the "new" builtin metrics: https://github.com/mlflow/mlflow/pull/9439#discussion_r1310759860

Reply:
Does this work if the type hint is `dict[str, MetricValue]`?

Reply:
What if `from __future__ import annotations` is imported in the code? Does this approach still work properly?

Reply:
It works with `from __future__ import annotations`,` dict[str, MetricValue]` gives `TypeError: 'type' object is not subscriptable`

Reply:
`dict[str, MetricValue]` is more common in newer versions of python. Can we support it?